File: iOS/TUIKit/TUICallKit/TUICallKit/UI/Recents/TUICallRecordCallsViewController.m
//
//  TUICallRecordCallsViewController.m
//  TUICallKit
//
//  Created by noah on 2023/2/27.
//  Copyright © 2023 Tencent. All rights reserved.
//

#import "TUICallRecordCallsViewController.h"
#import "Masonry.h"
#import "TUICore.h"
#import "TUICallingCommon.h"
#import "TUICallRecordCallsViewModel.h"
#import "TUICallRecordCallsCell.h"
#import "CallingLocalized.h"
#import "TUIThemeManager.h"

static NSString *const gRecordCallsViewModelKVOKeyPath = @"dataSource";

@interface TUICallRecordCallsViewController ()<UITableViewDelegate, UITableViewDataSource>

/// 头部导航容器视图
@property (nonatomic, strong) UIView *containerView;
/// 开始编辑按钮
@property (nonatomic, strong) UIButton *editButton;
/// 清除所有记录按钮
@property (nonatomic, strong) UIButton *clearButton;
/// 编辑完成按钮
@property (nonatomic, strong) UIButton *doneButton;
/// 分段控制视图
@property (nonatomic, strong) UISegmentedControl *segmentedControl;
/// 无最近通话提示文本
@property (nonatomic, strong) UILabel *notRecordCallsLabel;
/// 列表表头视图
@property (nonatomic, strong) UIView *tableHeaderView;
/// 通话记录列表
@property (nonatomic, strong) UITableView *recordCallsList;

@property (nonatomic, readwrite, strong) TUICallRecordCallsViewModel *recordCallsViewModel;

@end

@implementation TUICallRecordCallsViewController

- (instancetype)initWithRecordCallsUIStyle:(TUICallKitRecordCallsUIStyle)recordCallsUIStyle {
    self = [super init];
    if (self) {
        _recordCallsViewModel = [[TUICallRecordCallsViewModel alloc] init];
        _recordCallsViewModel.recordCallsUIStyle = recordCallsUIStyle;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = TUICallKitDynamicColor(@"callkit_recents_bg_color", @"#FFFFFF");
    [self constructViewHierarchy];
    [self activateConstraints];
    [self bindInteraction];
    [self bindViewModel];
    [self.recordCallsViewModel queryRecentCalls];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.recordCallsViewModel queryRecentCalls];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(windowDidChange)
                                                 name:UIWindowDidBecomeKeyNotification
                                               object:nil];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIWindowDidBecomeKeyNotification object:nil];
}

- (void)dealloc {
    [self removeBindViewModel];
}

- (void)constructViewHierarchy {
    [self.view addSubview:self.containerView];
    [self.containerView addSubview:self.segmentedControl];
    [self.containerView addSubview:self.editButton];
    [self.containerView addSubview:self.clearButton];
    [self.containerView addSubview:self.doneButton];
    [self.view addSubview:self.notRecordCallsLabel];
    [self.view addSubview:self.recordCallsList];
}

- (void)activateConstraints {
    CGFloat statusBarHeight = [TUICallingCommon getStatusBarHeight];
    [self.containerView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.centerX.width.equalTo(self.view);
        make.height.equalTo(@(statusBarHeight + 44));
    }];
    [self.segmentedControl mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.containerView).offset(statusBarHeight + 6.0);
        make.centerX.equalTo(self.containerView);
        make.width.equalTo(@(180));
        make.height.equalTo(@(32));
    }];
    [self.editButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self.segmentedControl);
        make.trailing.equalTo(self.containerView).offset(-20);
        make.width.height.equalTo(@(32));
    }];
    [self.clearButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self.segmentedControl);
        make.leading.equalTo(self.containerView).offset(20);
        make.height.equalTo(@(32));
    }];
    [self.doneButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self.segmentedControl);
        make.trailing.equalTo(self.containerView).offset(-20);
        make.height.equalTo(@(32));
    }];
    [self.recordCallsList mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.view).offset(statusBarHeight + 44);
        make.centerX.width.bottom.equalTo(self.view);
    }];
    [self.notRecordCallsLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.center.width.equalTo(self.view);
    }];
}

- (void)bindInteraction {
    [self.clearButton addTarget:self action:@selector(clearButtonClick:) forControlEvents:UIControlEventTouchUpInside];
    [self.editButton addTarget:self action:@selector(editButtonClick:) forControlEvents:UIControlEventTouchUpInside];
    [self.doneButton addTarget:self action:@selector(doneButtonClick:) forControlEvents:UIControlEventTouchUpInside];
    [self.segmentedControl addTarget:self action:@selector(segmentSelectItem:) forControlEvents:UIControlEventValueChanged];
}

- (void)bindViewModel {
    [self.recordCallsViewModel addObserver:self forKeyPath:gRecordCallsViewModelKVOKeyPath options:NSKeyValueObservingOptionNew context:nil];
}

- (void)removeBindViewModel {
    @try {
        [self.recordCallsViewModel removeObserver:self forKeyPath:gRecordCallsViewModelKVOKeyPath context:nil];
    }
    @catch (NSException *exception) {
    }
}

#pragma mark - event action

- (void)windowDidChange {
    [self.recordCallsViewModel queryRecentCalls];
}

- (void)clearButtonClick:(UIButton *)button {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:nil
                                                                             message:nil
                                                                      preferredStyle:UIAlertControllerStyleActionSheet];
    __weak typeof(self)weakSelf = self;
    UIAlertAction *clearAction = [UIAlertAction actionWithTitle:TUICallingLocalize(@"TUICallKit.Recents.clear.all")
                                                          style:UIAlertActionStyleDestructive
                                                        handler:^(UIAlertAction * _Nonnull action) {
        __strong typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf.recordCallsList setEditing:NO animated:YES];
        [strongSelf showButtonStatus:NO];
        [strongSelf.recordCallsViewModel clearAllRecordCalls];
    }];
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:TUICallingLocalize(@"TUICallKit.Recents.clear.cancel")
                                                           style:UIAlertActionStyleCancel
                                                         handler:^(UIAlertAction * _Nonnull action) {
    }];
    [alertController addAction:clearAction];
    [alertController addAction:cancelAction];
    [self presentViewController:alertController animated:YES completion:nil];
}

- (void)editButtonClick:(UIButton *)button {
    [self.recordCallsList setEditing:YES animated:YES];
    [self showButtonStatus:YES];
}

- (void)doneButtonClick:(UIButton *)button {
    [self.recordCallsList setEditing:NO animated:YES];
    [self showButtonStatus:NO];
}

- (void)callButtonClick:(UIButton *)button {
    
}

- (void)segmentSelectItem:(UISegmentedControl *)sender {
    [self showButtonStatus:NO];
    [self.recordCallsList setEditing:NO animated:YES];
    TUICallRecordCallsType recordCallsType = TUICallRecordCallsTypeAll;
    
    if (sender.selectedSegmentIndex == 1) {
        recordCallsType = TUICallRecordCallsTypeMissed;
    }
    
    [self.recordCallsViewModel switchRecordCallsType:recordCallsType];
}

- (void)showButtonStatus:(BOOL)isEdit {
    self.editButton.hidden = isEdit;
    self.clearButton.hidden = !isEdit;
    self.doneButton.hidden = !isEdit;
}

#pragma mark - observeValueForKeyPath

- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary<NSKeyValueChangeKey,id> *)change
                       context:(void *)context {
    if ([keyPath isEqualToString:gRecordCallsViewModelKVOKeyPath]) {
        dispatch_callkit_main_async_safe(^{
            [self.recordCallsList reloadData];
            
            if (self.recordCallsViewModel.dataSource <= 0) {
                [self.recordCallsList setEditing:NO animated:YES];
                [self showButtonStatus:NO];
            }
        });
    }
}

#pragma mark - UITableViewDataSource and UITableViewDelegate

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath {
    return UITableViewCellEditingStyleDelete;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        [self.recordCallsViewModel deleteRecordCall:indexPath];
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    }
}

- (NSString *)tableView:(UITableView *)tableView titleForDeleteConfirmationButtonForRowAtIndexPath:(NSIndexPath *)indexPath {
    return TUICallingLocalize(@"TUICallKit.Recents.delete");
}

- (nullable NSArray<UITableViewRowAction *> *)tableView:(UITableView *)tableView editActionsForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewRowAction *delete = [UITableViewRowAction rowActionWithStyle:UITableViewRowActionStyleDestructive
                                                                      title:TUICallingLocalize(@"TUICallKit.Recents.delete")
                                                                    handler:^(UITableViewRowAction * _Nonnull action,
                                                                              NSIndexPath * _Nonnull indexPath) {
        [self.recordCallsViewModel deleteRecordCall:indexPath];
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    }];
    delete.backgroundColor = [UIColor redColor];
    return @[delete];
}

- (UISwipeActionsConfiguration *)tableView:(UITableView *)tableView
trailingSwipeActionsConfigurationForRowAtIndexPath:(NSIndexPath *)indexPath  API_AVAILABLE(ios(11.0)){
    if (@available(iOS 11.0, *)) {
        UIContextualAction *deleteRowAction = [UIContextualAction contextualActionWithStyle:UIContextualActionStyleNormal
                                                                                      title:TUICallingLocalize(@"TUICallKit.Recents.delete")
                                                                                    handler:^(UIContextualAction * _Nonnull action,
                                                                                              __kindof UIView * _Nonnull sourceView,
                                                                                              void (^ _Nonnull completionHandler)(BOOL)) {
            [self.recordCallsViewModel deleteRecordCall:indexPath];
            [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
        }];
        deleteRowAction.backgroundColor = [UIColor redColor];
        UISwipeActionsConfiguration *config = [UISwipeActionsConfiguration configurationWithActions:@[deleteRowAction]];
        config.performsFirstActionWithFullSwipe = false;
        [self showButtonStatus:YES];
        return config;
    } else {
        return nil;
    }
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.recordCallsViewModel.dataSource ? self.recordCallsViewModel.dataSource.count : 1;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 64.0f;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    TUICallRecordCallsCell *cell = [tableView dequeueReusableCellWithIdentifier:NSStringFromClass([TUICallRecordCallsCell class])
                                                                   forIndexPath:indexPath];
    __weak typeof(self)weakSelf = self;
    cell.moreBtnClickedHandler = ^(TUICallRecordCallsCell *recordCallsCell) {
        __strong typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf.recordCallsViewModel jumpUserInfoController:indexPath navigationController:strongSelf.navigationController];
    };
    
    TUICallRecordCallsCellViewModel *viewModel = self.recordCallsViewModel.dataSource[indexPath.row];
    [cell bindViewModel:viewModel];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    [self.recordCallsViewModel repeatCall:indexPath];
}

#pragma mark - getter and setter

- (UIView *)containerView {
    if (!_containerView) {
        _containerView = [[UIView alloc] init];
        if (TUICallKitRecordCallsUIStyleClassic == self.recordCallsViewModel.recordCallsUIStyle) {
            _containerView.backgroundColor = TUICoreDynamicColor(@"head_bg_gradient_start_color", @"#EBF0F6");
        } else {
            _containerView.backgroundColor = TUICallKitDynamicColor(@"callkit_recents_bg_color", @"#FFFFFF");
        }
    }
    return _containerView;
}

- (UIButton *)clearButton {
    if (!_clearButton) {
        UIButton *clearButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [clearButton setTitle:TUICallingLocalize(@"TUICallKit.Recents.clear") forState:UIControlStateNormal];
        clearButton.titleLabel.textAlignment = isRTL() ? NSTextAlignmentRight : NSTextAlignmentLeft;
        clearButton.contentHorizontalAlignment = UIControlContentHorizontalAlignmentLeft;
        [clearButton setTitleColor:TUICallKitDynamicColor(@"callkit_nav_title_text_color", @"#000000") forState:UIControlStateNormal];
        clearButton.hidden = YES;
        [clearButton.titleLabel sizeToFit];
        clearButton.titleLabel.adjustsFontSizeToFitWidth = YES;
        _clearButton = clearButton;
    }
    return _clearButton;
}

- (UIButton *)editButton {
    if (!_editButton) {
        UIButton *editButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [editButton setImage:[TUICallingCommon getBundleImageWithName:@"ic_calls_edit"] forState:UIControlStateNormal];
        editButton.titleLabel.textAlignment = isRTL() ? NSTextAlignmentRight : NSTextAlignmentLeft;
        editButton.imageView.contentMode = UIViewContentModeScaleAspectFit;
        _editButton = editButton;
    }
    return _editButton;
}

- (UIButton *)doneButton {
    if (!_doneButton) {
        UIButton *doneButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [doneButton setTitle:TUICallingLocalize(@"TUICallKit.Recents.done") forState:UIControlStateNormal];
        doneButton.contentHorizontalAlignment = UIControlContentHorizontalAlignmentRight;
        [doneButton setTitleColor:TUICallKitDynamicColor(@"callkit_nav_title_text_color", @"#000000") forState:UIControlStateNormal];
        [doneButton.titleLabel sizeToFit];
        doneButton.titleLabel.adjustsFontSizeToFitWidth = YES;
        doneButton.hidden = YES;
        _doneButton = doneButton;
    }
    return _doneButton;
}

- (UISegmentedControl *)segmentedControl {
    if (!_segmentedControl) {
        NSArray *array = [NSArray arrayWithObjects:TUICallingLocalize(@"TUICallKit.Recents.all"),
                          TUICallingLocalize(@"TUICallKit.Recents.missed"), nil];
        UISegmentedControl *segment = [[UISegmentedControl alloc] initWithItems:array];
        segment.selectedSegmentIndex = 0;
        _segmentedControl = segment;
    }
    return _segmentedControl;
}

- (UIView *)tableHeaderView {
    if (!_tableHeaderView) {
        UIView *tableHeaderView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, UIScreen.mainScreen.bounds.size.width, 80)];
        UILabel *label = [[UILabel alloc] init];
        label.textColor = TUICallKitDynamicColor(@"callkit_recents_tableHeader_title_text_color", @"#000000");
        label.text = TUICallingLocalize(@"TUICallKit.Recents.calls");
        label.font = [UIFont fontWithName:@"PingFangHK-Semibold" size:34];
        label.textAlignment = isRTL() ? NSTextAlignmentRight : NSTextAlignmentLeft;
        [tableHeaderView addSubview:label];
        [label mas_makeConstraints:^(MASConstraintMaker *make) {
            make.center.equalTo(tableHeaderView);
            make.leading.equalTo(tableHeaderView).offset(20);
        }];
        _tableHeaderView = tableHeaderView;
    }
    return _tableHeaderView;
}

- (UITableView *)recordCallsList {
    if (!_recordCallsList) {
        UITableView *tableView = [[UITableView alloc] init];
        tableView.tableHeaderView = self.tableHeaderView;
        tableView.delegate = self;
        tableView.dataSource = self;
        tableView.backgroundColor = self.view.backgroundColor;
        [tableView registerClass:[TUICallRecordCallsCell class] forCellReuseIdentifier:NSStringFromClass([TUICallRecordCallsCell class])];
        _recordCallsList = tableView;
    }
    return _recordCallsList;
}

- (UILabel *)notRecordCallsLabel {
    if (!_notRecordCallsLabel) {
        UILabel *label = [[UILabel alloc] initWithFrame:CGRectZero];
        label.textColor = UIColor.grayColor;
        label.font = [UIFont systemFontOfSize:20];
        label.textAlignment = NSTextAlignmentCenter;
        label.hidden = YES;
        _notRecordCallsLabel = label;
    }
    return _notRecordCallsLabel;
}

@end

File: iOS/TUIKit/TUICallKit/TUICallKit-Swift/View/Component/Recents/TUICallRecordCallsViewController.swift
//
//  TUICallRecordCallsViewController.swift
//  
//
//  Created by vincepzhang on 2023/8/28.
//

import Foundation
import UIKit
import SnapKit
import TUICore

class TUICallRecordCallsViewController: UIViewController {
    
    private let viewModel = TUICallRecordCallsViewModel()
    private let dataSourceObserver = Observer()
    
    // 头部导航容器视图
    private lazy var containerView: UIView = {
        let view = UIView()
        if viewModel.recordCallsUIStyle == .classic {
            view.backgroundColor = TUICoreDefineConvert.getTUICoreDynamicColor(colorKey: "head_bg_gradient_start_color",
                                                                               defaultHex:  "#EBF0F6")
        } else {
            view.backgroundColor = TUICoreDefineConvert.getTUICoreDynamicColor(colorKey: "callkit_recents_bg_color",
                                                                               defaultHex:  "#FFFFFF")
        }
        return view
    }()
    
    // 开始编辑按钮
    private lazy var editButton: UIButton = {
        let button = UIButton(type: .custom)
        button.setImage(TUICallKitCommon.getBundleImage(name: "ic_calls_edit"), for: .normal)
        button.titleLabel?.textAlignment = TUICoreDefineConvert.getIsRTL() ? .right : .left
        button.imageView?.contentMode = .scaleAspectFit
        return button
    }()
    
    // 清除所有记录按钮
    private lazy var clearButton: UIButton = {
        let button = UIButton(type: .custom)
        button.setTitle(TUICallKitLocalize(key: "TUICallKit.Recents.clear"), for: .normal)
        button.titleLabel?.textAlignment = TUICoreDefineConvert.getIsRTL() ? .right : .left
        button.contentHorizontalAlignment = TUICoreDefineConvert.getIsRTL() ? .right : .left
        button.setTitleColor(TUICoreDefineConvert.getTUICallKitDynamicColor(colorKey: "callkit_nav_title_text_color",
                                                                            defaultHex: "#000000"), for: .normal)
        button.isHidden = true
        button.titleLabel?.sizeToFit()
        button.titleLabel?.adjustsFontSizeToFitWidth = true
        return button
    }()
    
    // 编辑完成按钮
    private lazy var doneButton: UIButton = {
        let button = UIButton(type: .custom)
        button.setTitle(TUICallKitLocalize(key: "TUICallKit.Recents.done"), for: .normal)
        button.contentHorizontalAlignment = .right
        button.setTitleColor(TUICoreDefineConvert.getTUICallKitDynamicColor(colorKey: "callkit_nav_title_text_color",
                                                                            defaultHex: "#000000"), for: .normal)
        button.titleLabel?.sizeToFit()
        button.titleLabel?.adjustsFontSizeToFitWidth = true
        button.isHidden = true
        return button
    }()
    
    // 分段控制视图
    private lazy var segmentedControl: UISegmentedControl = {
        let items = [TUICallKitLocalize(key: "TUICallKit.Recents.all"), TUICallKitLocalize(key: "TUICallKit.Recents.missed")]
        let control = UISegmentedControl(items: items as [Any])
        control.selectedSegmentIndex = 0
        return control
    }()
    
    // 无最近通话提示文本
    private lazy var notRecordCallsLabel: UILabel = {
        let label = UILabel(frame: .zero)
        label.textColor = UIColor.gray
        label.font = UIFont.systemFont(ofSize: 20)
        label.textAlignment = .center
        label.isHidden = true
        return label
    }()
    
    // 列表表头视图
    private lazy var tableHeaderView: UIView = {
        let tableHeaderView = UIView(frame: CGRect(x: 0, y: 0, width: UIScreen.main.bounds.size.width, height: 80))
        let label = UILabel()
        label.textColor = TUICoreDefineConvert.getTUICallKitDynamicColor(colorKey: "callkit_recents_tableHeader_title_text_color",
                                                                         defaultHex: "#000000")
        label.text = TUICallKitLocalize(key: "TUICallKit.Recents.calls")
        label.font = UIFont(name: "PingFangHK-Semibold", size: 34)
        label.textAlignment = TUICoreDefineConvert.getIsRTL() ? .right : .left
        tableHeaderView.addSubview(label)
        label.snp.makeConstraints { make in
            make.center.equalToSuperview()
            make.leading.equalTo(tableHeaderView).offset(20)
        }
        return tableHeaderView
    }()
    
    // 通话记录列表
    private lazy var recordCallsList: UITableView = {
        let tableView = UITableView(frame: .zero, style: .plain)
        tableView.tableHeaderView = tableHeaderView
        tableView.backgroundColor = view.backgroundColor
        return tableView
    }()
    
    
    init(recordCallsUIStyle: TUICallKitRecordCallsUIStyle) {
        super.init(nibName: nil, bundle: nil)
        
        viewModel.recordCallsUIStyle = recordCallsUIStyle
        
        view.backgroundColor = TUICoreDefineConvert.getTUICallKitDynamicColor(colorKey: "callkit_recents_bg_color",
                                                                              defaultHex: "#FFFFFF")
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // MARK: Observe
    func registerObserve() {
        NotificationCenter.default.addObserver(self, selector: #selector(windowDidChange),
                                               name: UIWindow.didBecomeKeyNotification, object: nil)
        
        viewModel.dataSource.addObserver(dataSourceObserver) {[weak self] _, _ in
            guard let self = self else { return }
            DispatchQueue.main.async { [weak self] in
                guard let self = self else { return }
                self.recordCallsList.reloadData()
            }
        }
    }
    
    func unregisterObserve() {
        NotificationCenter.default.removeObserver(self, name: UIWindow.didBecomeKeyNotification, object: nil)
        viewModel.dataSource.removeObserver(dataSourceObserver)
    }
    
    // 页面出现的时候注册刷行UI的数据回调，页面消失关闭，防止页面UI刷新在非主线程中运行
    override func viewWillAppear(_ animated: Bool) {
        registerObserve()
        viewModel.queryRecentCalls()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        unregisterObserve()
    }
    
    //MARK: UI Specification Processing
    override func viewDidLoad() {
        super.viewDidLoad()
        constructViewHierarchy()
        activateConstraints()
        bindInteraction()
        viewModel.queryRecentCalls()
    }
    
    func constructViewHierarchy() {
        view.addSubview(containerView)
        containerView.addSubview(segmentedControl)
        containerView.addSubview(editButton)
        containerView.addSubview(clearButton)
        containerView.addSubview(doneButton)
        view.addSubview(notRecordCallsLabel)
        view.addSubview(recordCallsList)
    }
    
    func activateConstraints() {
        containerView.snp.makeConstraints { make in
            make.top.centerX.width.equalTo(view)
            make.height.equalTo(StatusBar_Height + 44)
        }
        segmentedControl.snp.makeConstraints { make in
            make.top.equalTo(containerView).offset(StatusBar_Height + 6.0)
            make.centerX.equalTo(containerView)
            make.width.equalTo(180)
            make.height.equalTo(32)
        }
        editButton.snp.makeConstraints { make in
            make.centerY.equalTo(segmentedControl)
            make.trailing.equalTo(containerView).offset(-20)
            make.width.height.equalTo(32)
        }
        clearButton.snp.makeConstraints { make in
            make.centerY.equalTo(segmentedControl)
            make.leading.equalTo(containerView).offset(20)
            make.height.equalTo(32)
        }
        doneButton.snp.makeConstraints { make in
            make.centerY.equalTo(segmentedControl)
            make.trailing.equalTo(containerView).offset(-20)
            make.height.equalTo(32)
        }
        recordCallsList.snp.makeConstraints { make in
            make.top.equalTo(view).offset(StatusBar_Height + 44)
            make.centerX.width.bottom.equalTo(view)
        }
        notRecordCallsLabel.snp.makeConstraints { make in
            make.center.width.equalTo(view)
        }
    }
    
    func bindInteraction() {
        recordCallsList.delegate = self
        recordCallsList.dataSource = self
        recordCallsList.register(TUICallRecordCallsCell.self, forCellReuseIdentifier: NSStringFromClass(TUICallRecordCallsCell.self))
        
        clearButton.addTarget(self, action: #selector(clearButtonClick(_:)), for: .touchUpInside)
        editButton.addTarget(self, action: #selector(editButtonClick(_:)), for: .touchUpInside)
        doneButton.addTarget(self, action: #selector(doneButtonClick(_:)), for: .touchUpInside)
        segmentedControl.addTarget(self, action: #selector(segmentSelectItem(_:)), for: .valueChanged)
    }
    
    @objc func windowDidChange() {
        viewModel.queryRecentCalls()
    }
    
    @objc func clearButtonClick(_ button: UIButton) {
        let alertController = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        let clearAction = UIAlertAction(title: TUICallKitLocalize(key: "TUICallKit.Recents.clear.all"),
                                        style: .destructive) { [weak self] _ in
            guard let self = self else { return }
            self.recordCallsList.setEditing(false, animated: true)
            self.setShowEditButton(true)
            self.viewModel.deleteAllRecordCalls()
        }
        let cancelAction = UIAlertAction(title: TUICallKitLocalize(key: "TUICallKit.Recents.clear.cancel"), style: .cancel)
        alertController.addAction(clearAction)
        alertController.addAction(cancelAction)
        present(alertController, animated: true, completion: nil)
    }
    
    @objc func editButtonClick(_ button: UIButton) {
        recordCallsList.setEditing(true, animated: true)
        setShowEditButton(false)
    }
    
    @objc func doneButtonClick(_ button: UIButton) {
        recordCallsList.setEditing(false, animated: true)
        setShowEditButton(true)
    }
    
    @objc func segmentSelectItem(_ sender: UISegmentedControl) {
        var type: TUICallRecordCallsType = .all
        
        if sender.selectedSegmentIndex == 1 {
            type = .missed
        }
        
        viewModel.switchRecordCallsType(type)
    }
    
    func setShowEditButton(_ isShow: Bool) {
        editButton.isHidden = !isShow
        clearButton.isHidden = isShow
        doneButton.isHidden = isShow
    }
}

extension TUICallRecordCallsViewController: UITableViewDataSource, UITableViewDelegate {
    func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return true
    }
    
    func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCell.EditingStyle {
        return .delete
    }
    
    @available(iOS 11.0, *)
    func tableView(_ tableView: UITableView, trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath) -> UISwipeActionsConfiguration? {
        let deleteRowAction = UIContextualAction(style: .normal,
                                                 title: TUICallKitLocalize(key: "TUICallKit.Recents.delete"))
        { [weak self] action, sourceView, completionHandler in
            guard let self = self else { return }
            self.viewModel.deleteRecordCall(indexPath)
        }
        deleteRowAction.backgroundColor = .red
        let config = UISwipeActionsConfiguration(actions: [deleteRowAction])
        config.performsFirstActionWithFullSwipe = false
        self.setShowEditButton(false)
        return config
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 64.0
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return viewModel.dataSource.value.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: NSStringFromClass(TUICallRecordCallsCell.self),
                                                       for: indexPath) as? TUICallRecordCallsCell else { return UITableViewCell() }
        cell.moreBtnClickedHandler =  { [weak self] in
            guard let self = self else { return }
            guard let navigationController = self.navigationController else { return }
            self.viewModel.jumpUserInfoController(indexPath: indexPath, navigationController: navigationController)
        }
        cell.configViewModel(viewModel.dataSource.value[indexPath.row])
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        viewModel.repeatCall(indexPath)
    }
}
